### PR TITLE
Avoid double deep_copy in create_layout_right_mirror_view_and_copy for compatible layouts

### PR DESCRIPTION
**Pull request by aprokop**
This drops the number of host-device copies in the distributed nearest neighbors to 2. In a benchmark I'm running, 98% of the time is taken by copies. So, this speeds things up by 30%. Just the first patch of several to improve the performance.

### CODE REVIEW

File: src/details/ArborX_DetailsUtils.hpp [hunk 1]
```diff
@@ -119,6 +119,10 @@ create_layout_right_mirror_view_and_copy(ExecutionSpace const &execution_space,
   {
     return src;
   }
+  else if constexpr (has_compatible_layout)
+  {
+    return Kokkos::create_mirror_view_and_copy(memory_space, src);
+  }
```
Comment (masterleinad) [inline src/details/ArborX_DetailsUtils.hpp:125]: Can't you do
```C++
  if constexpr (has_compatible_layout)
  {
    return Kokkos::create_mirror_view_and_copy(memory_space, src);
  }
```
? If the memory space is the same `create_mirror_view_and_copy` also just returns a (shallow) copy.

Comment (aprokop) [reply src/details/ArborX_DetailsUtils.hpp:125]: Good point. Wonder why we didn't do it.

**Review by masterleinad — APPROVED** [review]

**Comment by dalg24** [comment]
Did you consider adding any tool-based testing to make sure that we don't introduce regressions to that "two copies at most" behavior we desire?

**Comment by aprokop** [comment]
> Did you consider adding any tool-based testing to make sure that we don't introduce regressions to that "two copies at most" behavior we desire?

No. I don't know what that would be. There's a larger question whether we even need this routines. Without 2d/3d views, I think everything in the distributed section is going to be 1d views that we control. Given that the distributor is in details, it should probably follow it.

**Comment by dalg24** [comment]
You saw how we set callbacks in our tests to check that region an kernel names are prefixed. There are hooks for deep_copy as well and you could have your lambda capture a reference to a counter that you increment when the tool is called. Then you observe that count and make assertions based on the behavior you want to guarantee. Does that make sense ?

**Comment by aprokop** [comment]
> You saw how we set callbacks in our tests to check that region an kernel names are prefixed. There are hooks for deep_copy as well and you could have your lambda capture a reference to a counter that you increment when the tool is called. Then you observe that count and make assertions based on the behavior you want to guarantee. Does that make sense ?

Not doing it as I will get rid of using `create_layout_right` in Distributor anyway.

**Comment by aprokop** [comment]
CUDA builds did not run, but that's fine.

**Pull request merged**